src/node_worker.cc

```diff
@@ -114,16 +149,27 @@ class WorkerThreadData {
     : w_(w) {
     CHECK_EQ(uv_loop_init(&loop_), 0);
 
-    Isolate* isolate = NewIsolate(
-        w->array_buffer_allocator_.get(),
-        &loop_,
-        w->platform_);
-    CHECK_NOT_NULL(isolate);
+    Isolate::CreateParams params;
```

Comment:
For my own understanding...`WorkerThreadData` is actually closer to `NodeMainInstance` for the main thread even though we don't have an equivalent abstraction of `Worker` for the main thread at the moment? (This was my impression when I tried with https://github.com/nodejs/node/issues/29925 locally but I decided to take some more time to figure out a proper class hierarchy before moving forward..)

I wonder whether it makes sense to have `Environment` points to a base class of `Worker` and `MainThread` (?) if they have a longer life time than `Environment`.

Reply:
> For my own understanding...`WorkerThreadData` is actually closer to `NodeMainInstance` for the main thread even though we don't have an equivalent abstraction of `Worker` for the main thread at the moment? (This was my impression when I tried with #29925 locally but I decided to take some more time to figure out a proper class hierarchy before moving forward..)

It is somewhat similar, yes.

> I wonder whether it makes sense to have `Environment` points to a base class of `Worker` and `MainThread` (?) if they have a longer life time than `Environment`.

What would we use that for? And how would that fit into the embedder API? Both Workers and the main thread have a lifetime longer than the Environment, but only very slightly so…

Reply:
For example, `Worker`, `NodeMainInstance` and `Environment` all keep copies of arguments and exec arguments. We could reduce the duplication by only keeping them in `Worker` and `MainThread`. There seem to be a lot of duplication along the hierarchy at the moment..

Reply:
I would be careful about that… deduplication is nice but right now moving towards a better embedder API is my primary goal, and linking from `Environment` to a superclass of `Worker` and `NodeMainInstance` sounds a bit fragile

Reply:
If anything...I think a public version of that super class could be a good base for a newer set of better embedder APIs? (but maybe it's just a crazy idea, I have not given too much thought into it).